vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/contextmenu/GridContextMenu.java

```diff
@@ -202,4 +207,58 @@ public Registration addGridContextMenuOpenedListener(
         return super.addOpenedChangeListener(ev -> listener.onComponentEvent(new GridContextMenuOpenedEvent<>(ev.getSource(), ev.isFromClient())));
     }
 
+    /**
+     * Gets the callback function that is executed before the context menu is
+     * opened.
+     *
+     * <p>
+     * The dynamic context renderer allows for customizing the contents of the
+     * context menu before it is open.
+     * </p>
+     *
+     * @return the callback function that is executed before opening the context
+     *         menu, or {@code null} if not specified.
+     */
+    public SerializableFunction<T, Boolean> getContextMenuDynamicContentRenderer() {
+        return contextMenuDynamicContentRenderer;
+    }
+
+    /**
+     * Sets a callback that is executed before the context menu is opened.
+     *
+     * <p>
+     * This callback receives the clicked item (if any) as an input parameter
+     * and further can dynamically modify the contents of the context menu. This
+     * is useful in situations where the context menu items cannot be known in
+     * advance and depend on the specific context (i.e. clicked row) and thus
+     * can be configured dynamically.
+     *
+     * The boolean return value of this callback specifies if the context menu
+     * will be opened.
+     * </p>
+     *
+     * @param contextMenuDynamicContentRenderer
+     *            the callback function that will be executed before opening the
+     *            context menu.
+     */
+    public void setContextMenuDynamicContentRenderer(
```

Comment:
I have some concerns about the naming of these new methods.

First of all, the "ContextMenu"-part in the method seems redundant, as the method is in the `ContextMenu`-class, so it's already clear where the content will be rendered. Also, the terms "Dynamic" and "Content" are kind of included in the renderer-concept. So IMO the method could as well be called `setRenderer` without losing any information.

But then we come to the bigger problem: this renderer-function is not conceptually a same kind of renderer as in other Java components in Vaadin. When using renderers, the component does the actual rendering for you, while the user can use a higher-level API, eg. to provide the components or template strings that should be rendered. Here the function is just a callback, which you can use to manually remove and add components to the menu, or do any other side-effects you want.

Actually, because of the `boolean` return value, the method should technically be called `setOpenableProvider` or `setEnabledProvider`, or something like that. The rendering is still just a side-effect that the user can choose to do or not.

But if we ignore that part, the function is basically a callback to execute before the opening. So IMO it should be an event listener. But it should still be possible to prevent the menu from opening based on the item, so the event could have a method for that. This is how I'd imagine the API to be used:
```java
contextMenu.addBeforeOpenListener(event -> {
  if (event.getItem() == null) {
    event.preventOpen();
  } else {
    // remove/add stuff to contextMenu
  }
});
```

Not the prettiest API, but I think it would be conceptually more correct. @Legioth, can you provide your opinion?

Reply:
The general convention in e.g. `Binder` is that there's an `addXyzListener` method in situations where multiple competing callbacks are supported and a `setXyzHandler` method if there can only be one. In this case, I think it makes most sense to only have one callback, which means that the "handler" concept might be appropriate.

Based on that point of view, my name suggestion without too much consideration would be `setDynamicContentHandler`. When removing "renderer" from the name, the words "dynamic" and "content" are again meaningful. I see the possibility of returning `false` more as an optimization rather than a primary feature, so I don't think the method name needs to take that into account.

Reply:
So do I understand you correctly:
- You agree that "renderer" is a misleading term in this context
- It's okay to name this method based on dynamic content, because that's the use case it's created for, although the method is technically just a generic callback which doesn't enforce the user to use it for dynamic content handling

Reply:
- Yes
- Yes, the method can be named based on what it is intended to be used for rather than all the ways it can potentially be abused.

Reply:
Choosing to not show a menu at all is really "dynamic". Furthermore, it's a special form of "content" in the same way that the value of a list variable can either be an empty list or `null`.

Reply:
Ok, that's fine for me. Thanks for your input!

So, @gatanaso, unless you disagree or want to discuss more, can you rename the methods to `setDynamicContentHandler` and `getDynamicContentHandler`?

Reply:
@pekam Yes, I will update the methods accordingly.